worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/BlockStateMask.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * WorldEdit, a Minecraft world manipulation toolkit
+ * Copyright (C) sk89q <http://www.sk89q.com>
+ * Copyright (C) WorldEdit team and contributors
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.function.mask;
+
+import com.sk89q.worldedit.extent.Extent;
+import com.sk89q.worldedit.math.BlockVector3;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+public class BlockStateMask extends AbstractExtentMask {
+
+    private final Map<String, String> states;
+
+    public BlockStateMask(Extent extent, Map<String, String> states) {
+        super(extent);
+        this.states = states;
+    }
+
+    @Override
+    public boolean test(BlockVector3 vector) {
+        Map<String, Object> existingStates = getExtent().getBlock(vector).getStates().entrySet().stream()
```

Comment:
I think this could use caching, it's going to be called often. Or perhaps there's another performant way to do it I can't think of.

Reply:
Realistically how often will it be called on a single position?

Perhaps do a "last position cache"

Reply:
pretty much never on the same position, since masks are generally used when iterating over regions.
is there a better way to do this? converting properties to strings is unfortunately kinda necessary since two different blocktypes have different property objects for the same name. we could cache properties per blocktype i suppose?

Reply:
so instead of checking a `Map<String, String>` each time, we could have a `Map<BlockType, Map<Property<>, Object>`, which allows us to do ```
BlockState block = getExtent().getBlock(vector);
cache.computeIfAbsent(block.getType(), ...).entrySet().stream().allMatch(entry -> block.getState(entry.getKey()) == entry.getValue());
```
that means only computing which properties to check once per block type, which will save a lot.